Google.Api.Gax.Grpc/Rest/PartialDecodingStreamReader.cs

```diff
@@ -0,0 +1,146 @@
+    {
+        _parsing = parsing;
+
+        var httpResponse = httpResponseTask.ConfigureAwait(false).GetAwaiter().GetResult();
+        // TODO [virost, 11/5]: should I wrap this exception?
+        httpResponse.EnsureSuccessStatusCode();
+        var stream = httpResponse.Content.ReadAsStreamAsync().ConfigureAwait(false).GetAwaiter().GetResult();
+        _streamReader = new StreamReader(stream);
+        
+        _readyResults = new Queue<TResponse>();
+        _currentBuffer = new StringBuilder();
+        _arrayClosed = false;
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    public async Task<bool> MoveNext(CancellationToken cancellationToken)
+    {
+        if (_readyResults.Count > 0)
+        {
+            Current = _readyResults.Dequeue();
+            return true;
+        }
+
+        while (_readyResults.Count == 0)
+        {
+            var buffer = new char[1000];
+            var taskRead = _streamReader.ReadAsync(buffer, 0, buffer.Length);
+            var cancellationTask = Task.Delay(-1, cancellationToken);
+            var resultTask = await Task.WhenAny(taskRead, cancellationTask).ConfigureAwait(false);
+
+            if (resultTask == cancellationTask)
+            {
+                // If the cancellationTask "wins" `Task.WhenAny` by being cancelled, the following await will throw TaskCancelledException.
+                await cancellationTask.ConfigureAwait(false);
+            }
+
+            var readLen = await taskRead.ConfigureAwait(false);
+            if (readLen == 0)
+            {
+                if (!_arrayClosed)
+                {
+                    var errorText = "Closing `]` bracket not received after iterating through the stream. " +
+                                    "This means that streaming ended without all objects transmitted. " +
+                                    "It is likely a result of server or network error.";
+                    throw new InvalidOperationException(errorText);
+                }
+
+                break;
+            }
+
+            var readChars = buffer.Take(readLen);
+            foreach (var c in readChars)
+            {
+                // Closing bracket for the top-level array
+                if (_currentBuffer.Length == 0 && c == ']')
+                {
+                    _arrayClosed = true;
+                    continue;
+                }
+
+                // Between-objects commas and spaces, as well as an opening bracket
+                // for the top-level array.
+                if (_currentBuffer.Length == 0 && c != '{')
+                {
+                    continue;
+                }
+                
+                _currentBuffer.Append(c);
+                if (c != '}')
+                {
+                    continue;
+                }
+
+                try
```

Comment:
I guess this will work for now... it feels like quite an expensive way of detecting the end of a message, where a JSON tokenizer could do this without having to try to parse and catch the exception. (We could have quite a few of these in a single message due to nesting, and it's a shame if a string of "}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}}" causes tens of exceptions to be thrown...

Reply:
(I would also like to remove our Json.NET dependency eventually - we currently only have it transitively due to Google.Apis.Auth; in a new Auth library we'd want to remove it. But that can be handled later.)

Reply:
I'm okay with this staying in for now, but:

- Please call `ToString` before this try/catch, so we can use the same string below.
- Please add a TODO to put a JSON tokenizer in here instead